### ISSUE TITLE
[Pinpoint] JVM 버전에 따른 ThreadLocalRandom 구현체 선택

### ISSUE THREAD

**Issue by lioolli** [naver/kaist-oss-course#64]
JDK 7에 추가된 ThreadLocalRandom은 싱글 쓰레드에서만 사용 가능하지만 기존 Random 구현체에 비해 빠른 성능을 가지고 있다. Pinpoint에서도 ThreadLocalRandom을 사용하는데, JDK6에는 이 클래스가 없기 때문에 Pinpoint 안에 JDK7의 ThreadLocalRandom 소스코드를 포함시켰다.

https://github.com/naver/pinpoint/blob/master/bootstrap-core/src/main/java/com/navercorp/pinpoint/bootstrap/util/jdk/ThreadLocalRandom.java

그런데 JDK8에서는 ThreadLocalRandom의 성능이 더욱 개선되었으나, 이 코드는 단순히 복사해서 사용할 수가 없다. 그래서 현재 실행중인 JDK의 버전에 따라 다음과 같이 어떤 ThreadLocalRandom을 사용할 것인지 선택하도록 수정한다.

| JVM 버전 | ThreadLocalRandom |
| --- | --- |
| 1.6 | com.pinpoint.bootstrap.util.jdk.ThreadLocalRandom |
| 1.7+ | java.util.concurrent.ThreadLocalRandom |

참고로, Pinpoint에서 ThreadLocalRandom을 사용하는 곳은 [SpanId](https://github.com/naver/pinpoint/blob/master/bootstrap-core/src/main/java/com/navercorp/pinpoint/bootstrap/context/SpanId.java) 뿐이다.

**Comment by rlaalstn5383** [comment]
제가 하겠습니다

**Comment by lsj1888** [comment]
@rlaalstn5383 진행해보시고, 잘 안되거나 궁금하신 내용있으면 이곳에 질문 올려주세요.

**Comment by rlaalstn5383** [comment]
SpanId를 jvm 1.6과 1.7+ 두 경우에 해당하는 두 개의 소스 파일로 만든 후 maven에서 취사 선택하도록 진행하면 될까요?

**Comment by lioolli** [comment]
@rlaalstn5383 아니요, pinpoint는 버전 별로 다르게 패키징 하지 않고 있고, 어떤 버전의 jvm에서 실행될 지는 런타임에서나 확인할 수 있기 때문에 빌드 결과물에 두 버전을 위한 클래스가 모두 포함되어 있어야 합니다.

SpanId를 추상화하는 것 보다는, ThreadLocalRandom을 인터페이스로 빼고, 팩토리로부터 구현체를 얻어서 사용하는데, 팩토리에서 객체를 생성할 때 jvm 버전에 따라서 구현체를 선택하는 식으로 생각하고 있었습니다.
